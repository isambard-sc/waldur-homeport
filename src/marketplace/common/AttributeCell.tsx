import { Check, X } from '@phosphor-icons/react';
import React from 'react';

import { FormattedHtml } from '@waldur/core/FormattedHtml';

import { Attribute } from '../types';

import { SecretField } from './SecretField';

interface AttributeCellProps {
  value: string | string[] | number | boolean | undefined;
  attr: Attribute;
}

export const AttributeCell: React.FC<AttributeCellProps> = ({
  attr,
  value,
}) => {
  switch (attr.type) {
    case 'list': {
      const titles = [];
      if (Array.isArray(value)) {
        value.forEach((key) => {
          const option = attr.options.find((item) => item.key === key);
          if (option) {
            titles.push(option.title);
          }
        });
      }
      return (
        <>
          {titles.map((item, index) => (
            <span key={index}>
              <Check weight="bold" />
              {` ${item}`}
              <br />
            </span>
          ))}
        </>
      );
    }

    case 'boolean': {
      return value === true ? (
        <Check weight="bold" className="text-info" />
      ) : (
        <X weight="bold" className="text-danger" />
      );
    }

    case 'choice': {
      const option = attr.options.find((item) => item.key === value);
      return <>{option ? option.title : 'N/A'}</>;
    }

    case 'password': {
      return (
        <>{typeof value === 'string' ? <SecretField value={value} /> : 'N/A'}</>
      );
    }

    case 'html':
      return <FormattedHtml html={value.toString()} />;

    default:
      return <>{value === undefined ? 'N/A' : value}</>;
  }
};
