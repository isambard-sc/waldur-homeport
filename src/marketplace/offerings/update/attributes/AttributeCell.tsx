import { Field } from 'redux-form';

import { AwesomeCheckbox } from '@waldur/core/AwesomeCheckbox';
import { FieldError } from '@waldur/form';
import { InputField } from '@waldur/form/InputField';

import { configAttrField } from './utils';

export const AttributeCell = ({ attribute }) => {
  if (attribute.type === 'boolean') {
    return (
      <Field
        name="value"
        component={(prop) => (
          <AwesomeCheckbox label={attribute.title} {...prop.input} />
        )}
      />
    );
  }
  const attr = configAttrField(attribute);
  return (
    <>
      <Field name="value" component={InputField} {...attr} />
      <Field
        name="value"
        {...attr}
        component={(fieldProps) => <FieldError error={fieldProps.meta.error} />}
      />
    </>
  );
};
