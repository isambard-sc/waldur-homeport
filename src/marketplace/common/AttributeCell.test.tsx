import { render } from '@testing-library/react';
import { describe, it, expect } from 'vitest';

import { AttributeCell } from './AttributeCell';
import {
  ListAttribute,
  ChoiceAttribute,
  BooleanAttribute,
  StringAttribute,
} from './AttributeCell.fixtures';

describe('AttributeCell', () => {
  it('renders list attribute', () => {
    const wrapper = render(
      <AttributeCell attr={ListAttribute} value={['E5-2650v3', 'E5-2680v3']} />,
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders choice attribute', () => {
    const wrapper = render(
      <AttributeCell attr={ChoiceAttribute} value="gpu_NVidia_K80" />,
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders invalid choice attribute', () => {
    const wrapper = render(
      <AttributeCell attr={ChoiceAttribute} value="invalid" />,
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders boolean value', () => {
    const wrapper = render(
      <AttributeCell attr={BooleanAttribute} value={true} />,
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders negative boolean value', () => {
    const wrapper = render(
      <AttributeCell attr={BooleanAttribute} value={false} />,
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders string attribute', () => {
    const wrapper = render(
      <AttributeCell attr={StringAttribute} value="/opt/" />,
    );
    expect(wrapper.container).toMatchSnapshot();
  });
});
