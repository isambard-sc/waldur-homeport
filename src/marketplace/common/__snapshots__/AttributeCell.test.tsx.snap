// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AttributeCell > renders boolean value 1`] = `
<div>
  <svg
    class="text-info"
    fill="currentColor"
    height="1em"
    viewBox="0 0 256 256"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M232.49,80.49l-128,128a12,12,0,0,1-17,0l-56-56a12,12,0,1,1,17-17L96,183,215.51,63.51a12,12,0,0,1,17,17Z"
    />
  </svg>
</div>
`;

exports[`AttributeCell > renders choice attribute 1`] = `
<div>
  NVidia K80
</div>
`;

exports[`AttributeCell > renders invalid choice attribute 1`] = `
<div>
  N/A
</div>
`;

exports[`AttributeCell > renders list attribute 1`] = `
<div>
  <span>
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 256 256"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M232.49,80.49l-128,128a12,12,0,0,1-17,0l-56-56a12,12,0,1,1,17-17L96,183,215.51,63.51a12,12,0,0,1,17,17Z"
      />
    </svg>
     Intel Xeon E5-2650v3
    <br />
  </span>
  <span>
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 256 256"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M232.49,80.49l-128,128a12,12,0,0,1-17,0l-56-56a12,12,0,1,1,17-17L96,183,215.51,63.51a12,12,0,0,1,17,17Z"
      />
    </svg>
     Intel Xeon E5-2680v3
    <br />
  </span>
</div>
`;

exports[`AttributeCell > renders negative boolean value 1`] = `
<div>
  <svg
    class="text-danger"
    fill="currentColor"
    height="1em"
    viewBox="0 0 256 256"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M208.49,191.51a12,12,0,0,1-17,17L128,145,64.49,208.49a12,12,0,0,1-17-17L111,128,47.51,64.49a12,12,0,0,1,17-17L128,111l63.51-63.52a12,12,0,0,1,17,17L145,128Z"
    />
  </svg>
</div>
`;

exports[`AttributeCell > renders string attribute 1`] = `
<div>
  /opt/
</div>
`;
